File: CS/dxSample/Form1.cs
using DevExpress.Utils.Animation;
using DevExpress.Utils.Helpers;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.WinExplorer.ViewInfo;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace dxSample {
    public partial class Form1 : XtraForm {
        public Form1() {
            InitializeComponent();
            var startPath = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
            var startNode = new NodeNavigationInfo(startPath);
            SetActivePath(startNode);
        }
        NodeNavigationInfo currentPath;
        NodeNavigationInfo CurrentNode {
            get { return currentPath; }
            set {
                if(CurrentNode == value)
                    return;
                currentPath = value;
                UpdateDataSource();
            }
        }
        Rectangle GridClientBounds { get { return new Rectangle(Point.Empty, this.gridControl1.Size); } }

        static readonly Size ImageSize = new Size(256, 256);
        void UpdateDataSource() {
            Cursor oldCursor = Cursor.Current;
            Cursor.Current = Cursors.WaitCursor;
            try {
                if(!string.IsNullOrEmpty(CurrentNode.Path))
                    this.gridControl1.DataSource = FileSystemHelper.GetFileSystemEntries(CurrentNode.Path, IconSizeType.ExtraLarge, ImageSize);
                else
                    this.gridControl1.DataSource = null;
                this.barEditItem1.EditValue = CurrentNode.Path;
            }
            finally {
                Cursor.Current = oldCursor;
            }
        }

        private void winExplorerView1_ItemKeyDown(object sender, DevExpress.XtraGrid.Views.WinExplorer.WinExplorerViewItemKeyEventArgs e) {
            if(this.transitionManager1.IsTransition)
                return;
            switch(e.KeyInfo.KeyCode) {
                case Keys.Enter:
                    OnItemClick(e.ItemInfo);
                    break;
                case Keys.Back:
                    RunZoomOutAnimation(() => { GoBackCore(); });
                    break;
            }
        }
        private void winExplorerView1_ItemDoubleClick(object sender, DevExpress.XtraGrid.Views.WinExplorer.WinExplorerViewItemDoubleClickEventArgs e) {
            OnItemClick(e.ItemInfo);
        }

        private void OnItemClick(WinExplorerItemViewInfo itemInfo) {
            if(itemInfo == null || itemInfo.Row.RowKey is FileEntry)
                return;
            var entry = (FileSystemEntry)itemInfo.Row.RowKey;
            var activeNode = new NodeNavigationInfo(entry.Path, itemInfo.Bounds);
            RunZoomInAnimation(activeNode, () => { SetActivePath(activeNode); });
        }

        private void OnBackButtonClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            if(this.transitionManager1.IsTransition)
                return;
            RunZoomOutAnimation(() => { GoBackCore(); });
        }

        private void OnForwardButtonClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            if(this.transitionManager1.IsTransition)
                return;
            int currentIndex = PathHistory.IndexOf(CurrentNode);
            if(currentIndex >= PathHistory.Count - 1)
                return;
            var activeNode = PathHistory[currentIndex + 1];
            RunZoomInAnimation(activeNode, () => { GoForward(); });
        }

        List<NodeNavigationInfo> PathHistory = new List<NodeNavigationInfo>();
        private void SetActivePath(NodeNavigationInfo node) {
            int currentIndex = PathHistory.IndexOf(CurrentNode);
            CurrentNode = node;
            if(currentIndex > -1) {
                int pathCount = PathHistory.Count;
                while(pathCount > currentIndex + 1) {
                    PathHistory.RemoveAt(pathCount - 1);
                    pathCount--;
                }
            }
            PathHistory.Add(CurrentNode);
        }
        private void GoBackCore() {
            int currentIndex = PathHistory.IndexOf(CurrentNode);
            if(currentIndex < 1)
                return;
            CurrentNode = PathHistory[currentIndex - 1];
        }
        private void GoForward() {
            int currentIndex = PathHistory.IndexOf(CurrentNode);
            if(currentIndex >= PathHistory.Count - 1)
                return;
            CurrentNode = PathHistory[currentIndex + 1];
        }

        private void OnBreadCrumbValueChanged(object sender, EventArgs e) {
            var breadCrumb = (BreadCrumbEdit)sender;
            var path = breadCrumb.EditValue as string;
            if(this.transitionManager1.IsTransition || path == null)
                return;
            var activeNode = new NodeNavigationInfo(path,
                new Rectangle(GridClientBounds.Width / 2, GridClientBounds.Height / 2, 1, 1));
            RunZoomInAnimation(activeNode, () => { SetActivePath(activeNode); });
        }

        void RunZoomInAnimation(NodeNavigationInfo activeNode, Action action) {
            var zoomTransition = this.transitionManager1.GetTransition<ZoomTransition>(this.gridControl1);
            this.transitionManager1.StartTransition(this.gridControl1);
            zoomTransition.ActiveSettings = new ZoomTransitionSettings() {
                SourceBounds = activeNode.SourceBounds, TargetBounds = GridClientBounds
            };
            action();
            this.transitionManager1.EndTransition();
        }
        void RunZoomOutAnimation(Action action) {
            var zoomTransition = this.transitionManager1.GetTransition<ZoomTransition>(this.gridControl1);
            this.transitionManager1.StartTransition(this.gridControl1);
            zoomTransition.ActiveSettings = new ZoomTransitionSettings() {
                Direction = ZoomTransitionDirection.ZoomOut, SourceBounds = GridClientBounds
            };
            action();
            this.transitionManager1.EndTransition();
        }
        #region BreadCrumb node generation
        private void repositoryItemBreadCrumbEdit1_ValidatePath(object sender, BreadCrumbValidatePathEventArgs e) {
            if(!FileSystemHelper.IsDirExists(e.Path)) {
                e.ValidationResult = BreadCrumbValidatePathResult.Cancel;
                return;
            }
            e.ValidationResult = BreadCrumbValidatePathResult.CreateNodes;
        }
        private void repositoryItemBreadCrumbEdit1_NewNodeAdding(object sender, BreadCrumbNewNodeAddingEventArgs e) {
            e.Node.PopulateOnDemand = true;
        }
        private void repositoryItemBreadCrumbEdit1_QueryChildNodes(object sender, BreadCrumbQueryChildNodesEventArgs e) {
            string dir = e.Node.Path;
            if(!FileSystemHelper.IsDirExists(dir))
                return;
            string[] subDirs = FileSystemHelper.GetSubFolders(dir);
            for(int i = 0; i < subDirs.Length; i++) {
                e.Node.ChildNodes.Add(CreateNode(subDirs[i]));
            }
        }
        BreadCrumbNode CreateNode(string path) {
            string folderName = FileSystemHelper.GetDirName(path);
            return new BreadCrumbNode(folderName, folderName, true);
        }
        #endregion
    }
    class NodeNavigationInfo {
        public NodeNavigationInfo(string path) : this(path, Rectangle.Empty) { }
        public NodeNavigationInfo(string path, Rectangle sourceBounds) {
            SourceBounds = sourceBounds;
            Path = path;
        }
        public Rectangle SourceBounds { get; private set; }
        public string Path { get; private set; }
    }
}

File: VB/dxSample/Form1.vb
Imports DevExpress.Utils.Animation
Imports DevExpress.Utils.Helpers
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.WinExplorer.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Namespace dxSample

    Public Partial Class Form1
        Inherits DevExpress.XtraEditors.XtraForm

        Public Sub New()
            Me.InitializeComponent()
            Dim startPath = System.Environment.GetFolderPath(System.Environment.SpecialFolder.Desktop)
            Dim startNode = New dxSample.NodeNavigationInfo(startPath)
            Me.SetActivePath(startNode)
        End Sub

        Private currentPath As dxSample.NodeNavigationInfo

        Private Property CurrentNode As NodeNavigationInfo
            Get
                Return Me.currentPath
            End Get

            Set(ByVal value As NodeNavigationInfo)
                If Me.CurrentNode Is value Then Return
                Me.currentPath = value
                Me.UpdateDataSource()
            End Set
        End Property

        Private ReadOnly Property GridClientBounds As Rectangle
            Get
                Return New System.Drawing.Rectangle(System.Drawing.Point.Empty, Me.gridControl1.Size)
            End Get
        End Property

        Private Shared ReadOnly ImageSize As System.Drawing.Size = New System.Drawing.Size(256, 256)

        Private Sub UpdateDataSource()
            Dim oldCursor As System.Windows.Forms.Cursor = System.Windows.Forms.Cursor.Current
            System.Windows.Forms.Cursor.Current = System.Windows.Forms.Cursors.WaitCursor
            Try
                If Not String.IsNullOrEmpty(Me.CurrentNode.Path) Then
                    Me.gridControl1.DataSource = DevExpress.Utils.Helpers.FileSystemHelper.GetFileSystemEntries(Me.CurrentNode.Path, DevExpress.Utils.Helpers.IconSizeType.ExtraLarge, dxSample.Form1.ImageSize)
                Else
                    Me.gridControl1.DataSource = Nothing
                End If

                Me.barEditItem1.EditValue = Me.CurrentNode.Path
            Finally
                System.Windows.Forms.Cursor.Current = oldCursor
            End Try
        End Sub

        Private Sub winExplorerView1_ItemKeyDown(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.WinExplorer.WinExplorerViewItemKeyEventArgs)
            If Me.transitionManager1.IsTransition Then Return
            Select Case e.KeyInfo.KeyCode
                Case System.Windows.Forms.Keys.Enter
                    Me.OnItemClick(e.ItemInfo)
                Case System.Windows.Forms.Keys.Back
                    Me.RunZoomOutAnimation(Sub() Me.GoBackCore())
            End Select
        End Sub

        Private Sub winExplorerView1_ItemDoubleClick(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.WinExplorer.WinExplorerViewItemDoubleClickEventArgs)
            Me.OnItemClick(e.ItemInfo)
        End Sub

        Private Sub OnItemClick(ByVal itemInfo As DevExpress.XtraGrid.Views.WinExplorer.ViewInfo.WinExplorerItemViewInfo)
            If itemInfo Is Nothing OrElse TypeOf itemInfo.Row.RowKey Is DevExpress.Utils.Helpers.FileEntry Then Return
            Dim entry = CType(itemInfo.Row.RowKey, DevExpress.Utils.Helpers.FileSystemEntry)
            Dim activeNode = New dxSample.NodeNavigationInfo(entry.Path, itemInfo.Bounds)
            Me.RunZoomInAnimation(activeNode, Sub() Me.SetActivePath(activeNode))
        End Sub

        Private Sub OnBackButtonClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            If Me.transitionManager1.IsTransition Then Return
            Me.RunZoomOutAnimation(Sub() Me.GoBackCore())
        End Sub

        Private Sub OnForwardButtonClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            If Me.transitionManager1.IsTransition Then Return
            Dim currentIndex As Integer = Me.PathHistory.IndexOf(Me.CurrentNode)
            If currentIndex >= Me.PathHistory.Count - 1 Then Return
            Dim activeNode = Me.PathHistory(currentIndex + 1)
            Me.RunZoomInAnimation(activeNode, Sub() Me.GoForward())
        End Sub

        Private PathHistory As System.Collections.Generic.List(Of dxSample.NodeNavigationInfo) = New System.Collections.Generic.List(Of dxSample.NodeNavigationInfo)()

        Private Sub SetActivePath(ByVal node As dxSample.NodeNavigationInfo)
            Dim currentIndex As Integer = Me.PathHistory.IndexOf(Me.CurrentNode)
            Me.CurrentNode = node
            If currentIndex > -1 Then
                Dim pathCount As Integer = Me.PathHistory.Count
                While pathCount > currentIndex + 1
                    Me.PathHistory.RemoveAt(pathCount - 1)
                    pathCount -= 1
                End While
            End If

            Me.PathHistory.Add(Me.CurrentNode)
        End Sub

        Private Sub GoBackCore()
            Dim currentIndex As Integer = Me.PathHistory.IndexOf(Me.CurrentNode)
            If currentIndex < 1 Then Return
            Me.CurrentNode = Me.PathHistory(currentIndex - 1)
        End Sub

        Private Sub GoForward()
            Dim currentIndex As Integer = Me.PathHistory.IndexOf(Me.CurrentNode)
            If currentIndex >= Me.PathHistory.Count - 1 Then Return
            Me.CurrentNode = Me.PathHistory(currentIndex + 1)
        End Sub

        Private Sub OnBreadCrumbValueChanged(ByVal sender As Object, ByVal e As System.EventArgs)
            Dim breadCrumb = CType(sender, DevExpress.XtraEditors.BreadCrumbEdit)
            Dim path = TryCast(breadCrumb.EditValue, String)
            If Me.transitionManager1.IsTransition OrElse Equals(path, Nothing) Then Return
            Dim activeNode = New dxSample.NodeNavigationInfo(path, New System.Drawing.Rectangle(Me.GridClientBounds.Width \ 2, Me.GridClientBounds.Height \ 2, 1, 1))
            Me.RunZoomInAnimation(activeNode, Sub() Me.SetActivePath(activeNode))
        End Sub

        Private Sub RunZoomInAnimation(ByVal activeNode As dxSample.NodeNavigationInfo, ByVal action As System.Action)
            Dim zoomTransition = Me.transitionManager1.GetTransition(Of DevExpress.Utils.Animation.ZoomTransition)(Me.gridControl1)
            Me.transitionManager1.StartTransition(Me.gridControl1)
            zoomTransition.ActiveSettings = New DevExpress.Utils.Animation.ZoomTransitionSettings() With {.SourceBounds = activeNode.SourceBounds, .TargetBounds = Me.GridClientBounds}
            action()
            Me.transitionManager1.EndTransition()
        End Sub

        Private Sub RunZoomOutAnimation(ByVal action As System.Action)
            Dim zoomTransition = Me.transitionManager1.GetTransition(Of DevExpress.Utils.Animation.ZoomTransition)(Me.gridControl1)
            Me.transitionManager1.StartTransition(Me.gridControl1)
            zoomTransition.ActiveSettings = New DevExpress.Utils.Animation.ZoomTransitionSettings() With {.Direction = DevExpress.Utils.Animation.ZoomTransitionDirection.ZoomOut, .SourceBounds = Me.GridClientBounds}
            action()
            Me.transitionManager1.EndTransition()
        End Sub

'#Region "BreadCrumb node generation"
        Private Sub repositoryItemBreadCrumbEdit1_ValidatePath(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.BreadCrumbValidatePathEventArgs)
            If Not DevExpress.Utils.Helpers.FileSystemHelper.IsDirExists(e.Path) Then
                e.ValidationResult = DevExpress.XtraEditors.BreadCrumbValidatePathResult.Cancel
                Return
            End If

            e.ValidationResult = DevExpress.XtraEditors.BreadCrumbValidatePathResult.CreateNodes
        End Sub

        Private Sub repositoryItemBreadCrumbEdit1_NewNodeAdding(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.BreadCrumbNewNodeAddingEventArgs)
            e.Node.PopulateOnDemand = True
        End Sub

        Private Sub repositoryItemBreadCrumbEdit1_QueryChildNodes(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.BreadCrumbQueryChildNodesEventArgs)
            Dim dir As String = e.Node.Path
            If Not DevExpress.Utils.Helpers.FileSystemHelper.IsDirExists(dir) Then Return
            Dim subDirs As String() = DevExpress.Utils.Helpers.FileSystemHelper.GetSubFolders(dir)
            For i As Integer = 0 To subDirs.Length - 1
                e.Node.ChildNodes.Add(Me.CreateNode(subDirs(i)))
            Next
        End Sub

        Private Function CreateNode(ByVal path As String) As BreadCrumbNode
            Dim folderName As String = DevExpress.Utils.Helpers.FileSystemHelper.GetDirName(path)
            Return New DevExpress.XtraEditors.BreadCrumbNode(folderName, folderName, True)
        End Function
'#End Region
    End Class

    Friend Class NodeNavigationInfo

        Private _SourceBounds As Rectangle, _Path As String

        Public Sub New(ByVal path As String)
            Me.New(path, System.Drawing.Rectangle.Empty)
        End Sub

        Public Sub New(ByVal path As String, ByVal sourceBounds As System.Drawing.Rectangle)
            Me.SourceBounds = sourceBounds
            Me.Path = path
        End Sub

        Public Property SourceBounds As Rectangle
            Get
                Return _SourceBounds
            End Get

            Private Set(ByVal value As Rectangle)
                _SourceBounds = value
            End Set
        End Property

        Public Property Path As String
            Get
                Return _Path
            End Get

            Private Set(ByVal value As String)
                _Path = value
            End Set
        End Property
    End Class
End Namespace
